kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/KubernetesCrudDispatcher.java

```diff
@@ -335,40 +317,44 @@ private String fetchResourceNameFromWatchRequestPath(String path) {
     return name.isEmpty()? null: name;
   }
 
-  private String fetchResource(String path) {
-    List<String> items = new ArrayList<>();
-    AttributeSet query = attributeExtractor.fromPath(path);
-
-    map.entrySet().stream()
-      .filter(entry -> entry.getKey().matches(query))
-      .forEach(entry -> items.add(entry.getValue()));
-
-    if (items.isEmpty()) {
-      return null;
-    } else if (items.size() == 1) {
-      return items.get(0);
-    } else {
-      return responseComposer.compose(items);
-    }
-  }
-
-
-  private int doDelete(String path, String event) {
-    List<AttributeSet> items = findItems(attributeExtractor.fromPath(path));
+  private int doDelete(String path) {
+    AttributeSet fromPath = attributeExtractor.fromPath(path);
+    List<AttributeSet> items = findItems(fromPath);
 
     if (items.isEmpty()) return HttpURLConnection.HTTP_NOT_FOUND;
 
+    items.forEach(item -> processEvent(path, fromPath, item, null));
+    return HttpURLConnection.HTTP_OK;
+  }
 
-    items.forEach(item -> {
-      if (event != null && !event.isEmpty()) {
-        watchEventListeners.stream()
-          .filter(listener -> listener.attributeMatches(item))
-          .forEach(listener -> listener.sendWebSocketResponse(map.get(item), event));
+  private void processEvent(String path, AttributeSet pathAttributes, AttributeSet oldAttributes, String newState) {
+    String existing = map.remove(oldAttributes);
+    AttributeSet newAttributes = null;
+    if (newState != null) {
+      newAttributes = kubernetesAttributesExtractor.fromResource(newState);
+      // corner case - we need to get the plural from the path
+      if (!newAttributes.containsKey(KubernetesAttributesExtractor.PLURAL)) {
+        newAttributes = AttributeSet.merge(pathAttributes, newAttributes);
       }
-      String existing = map.remove(item);
-      crdProcessor.process(path, existing, true);
-    });
-    return HttpURLConnection.HTTP_OK;
+      map.put(newAttributes, newState);
+    }
+    if (!Objects.equals(existing, newState)) {
+      AttributeSet finalAttributeSet = newAttributes;
+      watchEventListeners.stream()
+        .forEach(listener -> {
+          boolean matchesOld = oldAttributes != null && listener.attributeMatches(oldAttributes);
+          boolean matchesNew = finalAttributeSet != null && listener.attributeMatches(finalAttributeSet);
+          if (matchesOld && matchesNew) {
+            listener.sendWebSocketResponse(newState, MODIFIED);
+          } else if (matchesOld) {
+            listener.sendWebSocketResponse(existing, "DELETED");
```

Comment:
nit, maybe you should use constant here too for consistency

Reply:
I'll update things to reuse Watcher.Action